File: Palindrome Linked List LC-234.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
  def isPalindrome(self, head: ListNode) -> bool:
    
    slow=fast=head
    
    while fast and fast.next:
      slow=slow.next
      fast=fast.next.next
    
    stack = []
    
    while slow:
      stack.append(slow.val)
      slow=slow.next
    
    while stack:
      if head.val != stack.pop():
        return False
      head=head.next

    return True

  
# STACK

#   def isPalindrome(self, head: ListNode) -> bool:
    
#     slow=fast=head
    
#     while fast and fast.next:
#       slow=slow.next
#       fast=fast.next.next
    
#     stack = []
    
#     while slow:
#       stack.append(slow.val)
#       slow=slow.next
    
#     while stack:
#       if head.val != stack.pop():
#         return False
#       head=head.next

#     return True

File: Palindrome Linked List LC-234.cpp
// tc=o(n/2+n/2+n/2)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode *prev=NULL, *curr=head, *fast=head;
        while(fast && fast->next){
            curr=curr->next;
            fast=fast->next->next;
        }
        
        fast=curr;
        while(fast){
            fast=fast->next;
            curr->next=prev;
            prev=curr;
            curr=fast;
        }
        curr=head;
        while(prev){
            if(prev->val!=curr->val) return false;
            prev=prev->next;
            curr=curr->next;
        }
        return true;
    }
};